contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -236,6 +242,143 @@ exec_tsql_stmt(PLtsql_execstate *estate, PLtsql_stmt *stmt, PLtsql_stmt *save_es
 	return rc;
 }
 
+static int
+exec_stmt_kill(PLtsql_execstate *estate, PLtsql_stmt_kill *stmt)
+{
+	PGPROC *proc;    	
+	Oid	sysadmin_oid = get_role_oid("sysadmin", false);  /* We should really use BABELFISH_SYSADMIN in tds_int.h . */
+	int spid = -1;
+	Assert(stmt->spid);     
+	spid = stmt->spid;
+
+	if (pltsql_explain_only)
+	{
+		StringInfoData query;
+
+		initStringInfo(&query);
+		appendStringInfo(&query, "KILL ");
+		appendStringInfoString(&query, psprintf("%d", spid));
+		append_explain_info(NULL, query.data);
+		pfree(query.data);
+		return PLTSQL_RC_OK;
+	}
+
+	/* Do not allow to run KILL inside a transaction. */
+	if (IsTransactionBlockActive())
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+				errmsg("%s command cannot be used inside user transactions.", "KILL")));
+	}
+
+	/* Require that the user has 'sysadmin' role. */
+	if (!has_privs_of_role(GetSessionUserId(), sysadmin_oid)) 
+		{	       
+		ereport(ERROR,
+			(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				errmsg("User does not have permission to use the KILL statement")));
+	}
+
+	/*
+	 * Convert SPID string to int.
+	 * Although the T-SQL grammar only accepts a numeric value, and it must always be present, play it safe.
+	 */
+
+	if (spid == 0) 
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Session ID %d is not valid", spid)));            
+	}
+	/*
+	 * SPID value must be positive; the T-SQL grammar allows only a non-negative nuber to be specified.
+	 * Yet, play it safe and test for it.
+	 * A variable or expression is not allowed and caught in the parser.
+	 * All other variants of T-SQL KILL are not supported, this is caught in the parser.
+	 */
+	if (spid <= 0)
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Session ID %d is not valid", spid)));
+	}
```

Comment:
Missed this in first rev, why do we need a separate condition for this? Can we add it above and append the comment there?

Reply:
removed the preceding test & comment, those are leftovers from when the datatype was a char string.